index.js

```diff
@@ -2,67 +2,37 @@
 const got = require('got');
 const isPlainObj = require('is-plain-obj');
 
-function ghGot(path, opts) {
-	if (typeof path !== 'string') {
-		return Promise.reject(new TypeError(`Expected \`path\` to be a string, got ${typeof path}`));
-	}
-
-	const env = process.env;
-
-	opts = Object.assign({
+module.exports = got.create({
+	options: got.mergeOptions(got.defaults.options, {
 		json: true,
-		token: env.GITHUB_TOKEN,
-		endpoint: env.GITHUB_ENDPOINT ? env.GITHUB_ENDPOINT.replace(/[^/]$/, '$&/') : 'https://api.github.com/'
-	}, opts);
-
-	opts.headers = Object.assign({
-		accept: 'application/vnd.github.v3+json',
-		'user-agent': 'https://github.com/sindresorhus/gh-got'
-	}, opts.headers);
-
-	if (opts.token) {
-		opts.headers.authorization = `token ${opts.token}`;
-	}
-
-	// https://developer.github.com/v3/#http-verbs
-	if (opts.method && opts.method.toLowerCase() === 'put' && !opts.body) {
-		opts.headers['content-length'] = 0;
-	}
-
-	const url = /^https?/.test(path) ? path : opts.endpoint + path;
-
-	if (opts.stream) {
-		return got.stream(url, opts);
-	}
-
-	return got(url, opts).catch(err => {
-		if (err.response && isPlainObj(err.response.body)) {
-			err.name = 'GitHubError';
-			err.message = `${err.response.body.message} (${err.statusCode})`;
+		token: process.env.GITHUB_TOKEN,
+		baseUrl: process.env.GITHUB_ENDPOINT ? process.env.GITHUB_ENDPOINT.replace(/[^/]$/, '$&/') : 'https://api.github.com/',
+		headers: {
+			accept: 'application/vnd.github.v3+json',
+			'user-agent': 'https://github.com/sindresorhus/gh-got'
+		}
+	}),
+	methods: got.defaults.methods,
+	handler: (options, next) => {
+		if (options.token) {
+			options.headers.authorization = `token ${options.token}`;
 		}
 
-		throw err;
-	});
-}
-
-const helpers = [
-	'get',
-	'post',
-	'put',
-	'patch',
-	'head',
-	'delete'
-];
+		if (options.method && options.method.toLowerCase() === 'put' && !options.body) {
+			options.headers['content-length'] = 0;
+		}
 
-ghGot.stream = (url, opts) => ghGot(url, Object.assign({}, opts, {
-	json: false,
-	stream: true
-}));
+		if (options.stream) {
+			return next(options);
+		}
 
-for (const x of helpers) {
-	const method = x.toUpperCase();
-	ghGot[x] = (url, opts) => ghGot(url, Object.assign({}, opts, {method}));
-	ghGot.stream[x] = (url, opts) => ghGot.stream(url, Object.assign({}, opts, {method}));
-}
+		return next(options).catch(err => {
```

Comment:
Can you make the handler function async and use `await` and normal try/catch here?

Reply:
That's not currently possible, because `got` assumes this is a sync function. I'll send a PR.

Reply:
Nah. That breaks streams.

Reply:
I don't see how that breaks streams. Can you push what you changed?

Reply:
> I don't see how that would break streams.

If handler is async, it'll return a `Promise` and not a stream.

> Can you push what you changed?

Are you still sure?

Reply:
Oh, I had assumed `handler` could be an async function too, like the `beforeRequest` handler. I think the `handler` should be able to be an async function. I see many use-cases where you want to do some async stuff there.

Reply:
Yeah, that'd be very nice. But there's another problem: need to do a wrapper for `.cancel()` and `.on()`. We could accept `handler` also as an object containing two properties: `streamHandler` and `promiseHandler`.

It'd look like:
```js
// Your handler:
async function promiseHandler(options, next, wrap) {
	const promise = next(options);
	wrap(promise); // You always need to do this :(

	try {
		return await promise;
	} catch (err) {
		// Modify `err` here
		throw err;
	}
}

// create.js (got)
...
	const createWrapper = handler => promise => {
		handler.on = promise.on;
		handler.cancel = promise.cancel;
	};

	function got(url, options) {
		try {
			options = mergeOptions(defaults.options, options);
			const handler = defaults.handler(normalizeArguments(url, options, defaults), next, createWrapper(handler));
			return handler;
		} catch (error) {
			return Promise.reject(error);
		}
	}
...
```

A better solution would be to create a new hook: `onError`. You could modify that there.